### PR TITLE
[MNG-7129] - Changes in maven core required to support incremental mave…

### PR DESCRIPTION
**Pull request by AlexanderAshitkin**
Changes in maven core  to inject caching extension

CC @ansip

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MNG-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MNG-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

File: maven-core/src/main/java/org/apache/maven/plugin/MojoExecutionStrategy.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+public interface MojoExecutionStrategy
+{
+
+    void execute( MavenSession session, List<MojoExecution> mojoExecutions, Executor executor )
+            throws LifecycleExecutionException;
+
+    /**
+     * Implementors must perform real mojo execution
+     */
+    interface Executor
```
Comment (rfscholte) [inline maven-core/src/main/java/org/apache/maven/plugin/MojoExecutionStrategy.java:41]: I don't like the idea of subclass, you hardly see it in the Maven codebase. Would it be possible change its usage to `Consumer<MojoExecution>` ?

**Review by rfscholte — CHANGES_REQUESTED** [review]

File: maven-core/src/main/java/org/apache/maven/plugin/MojoExecutionStrategy.java (hunk 1 above)
Comment (andrey-petrenko-db) [reply maven-core/src/main/java/org/apache/maven/plugin/MojoExecutionStrategy.java:41]: do you mean java.util.function.Consumer? but it since Java 1.8 can we use classes/interfaces from 1.8?

Comment (andrey-petrenko-db) [reply maven-core/src/main/java/org/apache/maven/plugin/MojoExecutionStrategy.java:41]: regarding subclasses - is there any reason to not use them?
could be moved outside but I'm not sure about naming, the best name here for this interface is MojoExecutor but it is already used

Comment (rfscholte) [reply maven-core/src/main/java/org/apache/maven/plugin/MojoExecutionStrategy.java:41]: Maven 4 will require Java 8, so yes. And by using the Consumer, there's no need for the subclass anymore.

Comment (andrey-petrenko-db) [reply maven-core/src/main/java/org/apache/maven/plugin/MojoExecutionStrategy.java:41]: ah ok, got it
this branch was created from 3.8
btw would it be possible to merge these changes to 3.8 branch? looks like there is no breaking changes
this would allow us to use/test cache extension immediately and not await and then migrate our project to maven 4

Comment (rfscholte) [reply maven-core/src/main/java/org/apache/maven/plugin/MojoExecutionStrategy.java:41]: My opinion is that we should leave Maven 3 as is and just focus on Maven 4. This feature might attract more people, so putting it into Maven 4 makes more sense to me.

Comment (AlexanderAshitkin) [reply maven-core/src/main/java/org/apache/maven/plugin/MojoExecutionStrategy.java:41]: > Maven 4 will require Java 8, so yes. And by using the Consumer, there's no need for the subclass anymore.

Robert, what do you think of a bit opposite approach - to refactor and rename MojoExecutor. It is to big and could be split in 3 concerns - project preprocessing, list of mojos processing, idividual mojo execution? MojoExecutor could be renamed to Project(Executor|Builder), and MojoExecutor implements Executor could be extracted

Comment (AlexanderAshitkin) [reply maven-core/src/main/java/org/apache/maven/plugin/MojoExecutionStrategy.java:41]: > My opinion is that we should leave Maven 3 as is and just focus on Maven 4. This feature might attract more people, so putting it into Maven 4 makes more sense to me.

Robert it make total sense, but from practical needs the extension have to be tested on a stable maven version. Version 4 is in progress and could not be used in real project without upfront investements and also might be not stable yet. So for testing and verification reason need change in version 3 - otherwise there is no way to test extension on a real project

File: maven-core/src/main/java/org/apache/maven/plugin/DefaultMojoExecutionStrategy.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.maven.execution.MavenSession;
+import org.apache.maven.lifecycle.LifecycleExecutionException;
+import org.codehaus.plexus.component.annotations.Component;
```
Comment (andrey-petrenko-db) [inline maven-core/src/main/java/org/apache/maven/plugin/DefaultMojoExecutionStrategy.java:24]: unused?

File: maven-core/src/main/java/org/apache/maven/plugin/MojoExecutionStrategy.java (hunk 1 above)
Comment (rfscholte) [reply maven-core/src/main/java/org/apache/maven/plugin/MojoExecutionStrategy.java:41]: Have you tried Maven 4? At its current state I would expect that a Maven project can be built with both Maven 3 and Maven 4. A few warnings have turned into errors, so I would love to see which issues you hit when this feature is merged on the develop branch (aka Maven 4).

Comment (andrey-petrenko-db) [reply maven-core/src/main/java/org/apache/maven/plugin/MojoExecutionStrategy.java:41]: > My opinion is that we should leave Maven 3 as is and just focus on Maven 4. This feature might attract more people, so putting it into Maven 4 makes more sense to me.

Yes I didn't mean to develop extension in 3.8, I thought just to add these minimal changes to core 3.8 and 4
Then we can use our current extension draft (from 3.8) and test in against real projects (our project is about 750 modules with tricky configuration sometimes)
Our extension built with 3.8 maven could be left as a draft in our fork

To summarise our plan was:

1) merge minimal changes to 3.8 and 4
2) create new branch from MNG-7129-master and apply changes which we have done in our draft extension version (bug fixes, support to change projects versions like it is done by maven set plugin)
3) raise PR from our branch to MNG-7129-master

Comment (michael-o) [reply maven-core/src/main/java/org/apache/maven/plugin/MojoExecutionStrategy.java:41]: If at all in 3.x I see it in 3.9. I haven't though checked what needs to be changed in Core.

Comment (AlexanderAshitkin) [reply maven-core/src/main/java/org/apache/maven/plugin/MojoExecutionStrategy.java:41]: > Maven 4 will require Java 8, so yes. And by using the Consumer, there's no need for the subclass anymore.

consumer doesnt declare handled exceptions in api which makes it inconvenient to propagate LifecycleExecutionException. i made executor a functional interface and could extract it to a seprate class if that's preferable

File: maven-core/src/main/java/org/apache/maven/lifecycle/internal/MojoExecutor.java [hunk 3]
```diff
@@ -140,21 +145,22 @@ else if ( Artifact.SCOPE_TEST.equals( classpath ) )
...
+    public void execute( final MavenSession session,
+                         List<MojoExecution> mojoExecutions,
+                         final ProjectIndex projectIndex ) throws LifecycleExecutionException
 
     {
-        DependencyContext dependencyContext = newDependencyContext( session, mojoExecutions );
+        final DependencyContext dependencyContext = newDependencyContext( session, mojoExecutions );
 
-        PhaseRecorder phaseRecorder = new PhaseRecorder( session.getCurrentProject() );
+        final PhaseRecorder phaseRecorder = new PhaseRecorder( session.getCurrentProject() );
```
Comment (andrey-petrenko-db) [inline maven-core/src/main/java/org/apache/maven/lifecycle/internal/MojoExecutor.java:155]: since java8 no need to make them final

File: maven-core/src/main/java/org/apache/maven/plugin/DefaultMojosExecutionStrategy.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.List;
+import java.util.function.Consumer;
+
+/**
+ * Default mojo execution strategy.
+ * It just iterates over mojoExecutions list and pass all of them to executor
+ */
+@Named
+@Singleton
+public class DefaultMojosExecutionStrategy implements MojosExecutionStrategy
```
Comment (rfscholte) [inline maven-core/src/main/java/org/apache/maven/plugin/DefaultMojosExecutionStrategy.java:36]: Is this the class where you want to have a different implementation for the cache.
I assume all mojos use the same executor, so it is possible to inject the Executor via the constructor?

Comment (andrey-petrenko-db) [reply maven-core/src/main/java/org/apache/maven/plugin/DefaultMojosExecutionStrategy.java:36]: > Is this the class where you want to have a different implementation for the cache.
> I assume all mojos use the same executor, so it is possible to inject the Executor via the constructor?

DefaultMojosExecutionStrategy and Executor in our case have different scopes
First singleton (maven session), second one has project scope

If make both of them singleton then it is required to pass a lot of parameters to execute method (PhaseRecorder, ProjectIndex etc...) most of them related to internal implementation details I'm not sure whether it is ok to make them public API

Comment (andrey-petrenko-db) [reply maven-core/src/main/java/org/apache/maven/plugin/DefaultMojosExecutionStrategy.java:36]: In other words each call of 
``` 
public void execute( final MavenSession session,
                         List<MojoExecution> mojoExecutions,
                         final ProjectIndex projectIndex ) throws LifecycleExecutionException
```

creates new Executor with the full context related to particular project
this allows us to not pass details between interfaces on api level

**Comment by gnodet** [comment]
This looks similar to https://github.com/apache/maven/pull/607/commits/0a6f1d3d2e864b1eb0d84e3bbb929560aba6d483

**Comment by maximilian-novikov-db** [comment]
> This looks roughly similar to [0a6f1d3](https://github.com/apache/maven/commit/0a6f1d3d2e864b1eb0d84e3bbb929560aba6d483) The https://github.com/apache/maven/tree/MNG-7129-master already contains much more so if the goal is to merge into maven 4 only, I'd suggest looking at this branch instead.

IMO the change in this pr is more localized that will provide better supportability (if it will be moved to an extension). We'd like to raise the same code change to 3.8.x

**Comment by maximilian-novikov-db** [comment]
> This looks roughly similar to [0a6f1d3](https://github.com/apache/maven/commit/0a6f1d3d2e864b1eb0d84e3bbb929560aba6d483) The https://github.com/apache/maven/tree/MNG-7129-master already contains much more so if the goal is to merge into maven 4 only, I'd suggest looking at this branch instead.

@gnodet our plan is the following:
1. merge the minimal changes to 3.x(@michael-o 3.9 ?) and 4, as it's a very localized change it it will provide better supportability if the caching module will be moved to an extension
2. create new branch from MNG-7129-master and apply changes which we have done in our draft extension version (bug fixes, support to change projects versions like it is done by maven set plugin)
3. raise PR from our branch to MNG-7129-master

**Comment by michael-o** [comment]
@maximilian-novikov-db Since you are using Java 8 features here, it cannot be merged to 3.8.x for sure. I need to go through these changes and try to understand them.

**Comment by andrey-petrenko-db** [comment]
> @maximilian-novikov-db Since you are using Java 8 features here, it cannot be merged to 3.8.x for sure. I need to go through these changes and try to understand them.

@michael-o We can raise PR to 3.8 with similar changes, initial changes here were made against 3.8 branch (Used java 1.7 features only, used annotations from plexus: org.codehaus.plexus.component.annotations.Component, org.codehaus.plexus.component.annotations.Requirement)

**Comment by michael-o** [comment]
> 
> 
> > @maximilian-novikov-db Since you are using Java 8 features here, it cannot be merged to 3.8.x for sure. I need to go through these changes and try to understand them.
> 
> @michael-o We can raise PR to 3.8 with similar changes, initial changes here were made against 3.8 branch (Used java 1.7 features only, used annotations from plexus: org.codehaus.plexus.component.annotations.Component, org.codehaus.plexus.component.annotations.Requirement) and now this PR reworked and there is used java 8 and maven 4 features

Let's stick with Java 8 and 3.9. The code should not diverge. This will make life for us and you easier.

**Comment by andrey-petrenko-db** [comment]
> > > @maximilian-novikov-db Since you are using Java 8 features here, it cannot be merged to 3.8.x for sure. I need to go through these changes and try to understand them.
> > 
> > 
> > @michael-o We can raise PR to 3.8 with similar changes, initial changes here were made against 3.8 branch (Used java 1.7 features only, used annotations from plexus: org.codehaus.plexus.component.annotations.Component, org.codehaus.plexus.component.annotations.Requirement) and now this PR reworked and there is used java 8 and maven 4 features
> 
> Let's stick with Java 8 and 3.9. The code should not diverge. This will make life for us and you easier.

Ok but is there a separate branch for 3.9 
Or 3.9 is planned to release from the master hence 3.9 equals 4.0?

**Comment by andrey-petrenko-db** [comment]
> > > > @maximilian-novikov-db Since you are using Java 8 features here, it cannot be merged to 3.8.x for sure. I need to go through these changes and try to understand them.
> > > 
> > > 
> > > @michael-o We can raise PR to 3.8 with similar changes, initial changes here were made against 3.8 branch (Used java 1.7 features only, used annotations from plexus: org.codehaus.plexus.component.annotations.Component, org.codehaus.plexus.component.annotations.Requirement) and now this PR reworked and there is used java 8 and maven 4 features
> > 
> > 
> > Let's stick with Java 8 and 3.9. The code should not diverge. This will make life for us and you easier.
> 
> Ok but is there a separate branch for 3.9 Or 3.9 is planned to release from the master hence 3.9 equals 4.0?

Just to clarify: we don't have a problem with a version itself. It does not matter 3.9 or 4.0. The problem with compatibility.
If we build maven from master with 4.0-SNAPSHOT and try to build out project on this version our build stuck. (Max checked this, however he did this long time ago)

- Currently we have 3.6 used for our prod
- We tried 3.8 - it works
- built snapshot from master and tried to build our project - fail, so it seems it requires migration to a new maven version - this is what we want to prevent to not block extension development till maven 4 stable version released

**Comment by michael-o** [comment]
> 
> 
> > > > @maximilian-novikov-db Since you are using Java 8 features here, it cannot be merged to 3.8.x for sure. I need to go through these changes and try to understand them.
> > > 
> > > 
> > > @michael-o We can raise PR to 3.8 with similar changes, initial changes here were made against 3.8 branch (Used java 1.7 features only, used annotations from plexus: org.codehaus.plexus.component.annotations.Component, org.codehaus.plexus.component.annotations.Requirement) and now this PR reworked and there is used java 8 and maven 4 features
> > 
> > 
> > Let's stick with Java 8 and 3.9. The code should not diverge. This will make life for us and you easier.
> 
> Ok but is there a separate branch for 3.9 Or 3.9 is planned to release from the master hence 3.9 equals 4.0?

Not yet. There is [this](https://github.com/apache/maven/tree/maven-3.8.x-resolver-1.7.x) which is the tentative 3.9.x branch. I will branch off a couple of weeks after 3.8.4 release. Something around 2021-12-10 or so.

**Comment by gnodet** [comment]
> > > > > @maximilian-novikov-db Since you are using Java 8 features here, it cannot be merged to 3.8.x for sure. I need to go through these changes and try to understand them.
> > > > 
> > > > 
> > > > @michael-o We can raise PR to 3.8 with similar changes, initial changes here were made against 3.8 branch (Used java 1.7 features only, used annotations from plexus: org.codehaus.plexus.component.annotations.Component, org.codehaus.plexus.component.annotations.Requirement) and now this PR reworked and there is used java 8 and maven 4 features
> > > 
> > > 
> > > Let's stick with Java 8 and 3.9. The code should not diverge. This will make life for us and you easier.
> > 
> > 
> > Ok but is there a separate branch for 3.9 Or 3.9 is planned to release from the master hence 3.9 equals 4.0?
> 
> Not yet. There is [this](https://github.com/apache/maven/tree/maven-3.8.x-resolver-1.7.x) which is the tentative 3.9.x branch. I will branch off a couple of weeks after 3.8.4 release. Something around 2021-12-10 or so.

My understanding was that the goal was to focus on 4.0.x.  The work has already stalled a bit and I'm not sure creating additional branches will help in any way.

**Comment by gnodet** [comment]
> > > > > @maximilian-novikov-db Since you are using Java 8 features here, it cannot be merged to 3.8.x for sure. I need to go through these changes and try to understand them.
> > > > 
> > > > 
> > > > @michael-o We can raise PR to 3.8 with similar changes, initial changes here were made against 3.8 branch (Used java 1.7 features only, used annotations from plexus: org.codehaus.plexus.component.annotations.Component, org.codehaus.plexus.component.annotations.Requirement) and now this PR reworked and there is used java 8 and maven 4 features
> > > 
> > > 
> > > Let's stick with Java 8 and 3.9. The code should not diverge. This will make life for us and you easier.
> > 
> > 
> > Ok but is there a separate branch for 3.9 Or 3.9 is planned to release from the master hence 3.9 equals 4.0?
> 
> Just to clarify: we don't have a problem with a version itself. It does not matter 3.9 or 4.0. The problem with compatibility. If we build maven from master with 4.0-SNAPSHOT and try to build our project then our build stuck. (Max checked this, however he did this long time ago)
> 
> * Currently we have 3.6 used for our prod
> * We tried 3.8 - it works
> * built snapshot from master and tried to build our project - fail, so it seems it requires migration to a new maven version - this is what we want to prevent to not block extension development till maven 4 stable version released

It would be nice to have reports of things that fail so that we can actually fix those.

**Comment by gnodet** [comment]
> > This looks roughly similar to [0a6f1d3](https://github.com/apache/maven/commit/0a6f1d3d2e864b1eb0d84e3bbb929560aba6d483) The https://github.com/apache/maven/tree/MNG-7129-master already contains much more so if the goal is to merge into maven 4 only, I'd suggest looking at this branch instead.
> 
> @gnodet our plan is the following:
> 
> 1. merge the minimal changes to 3.x(@michael-o to 3.9 ?) and 4, as it's a very localized change it it will provide better supportability if the caching module will be moved to an extension

Yes, it's similar to the commit I pointed at but done in a different way (i.e. introducing a `MojoExecutorStrategy` interface instead of a `IMojoExecutor` one). 

> 2. create new branch from [MNG-7129](https://issues.apache.org/jira/browse/MNG-7129)-master and apply changes which we have done in our draft extension version (bug fixes, support to change projects versions like it is done by maven set plugin)
> 3. raise PR from our branch to [MNG-7129](https://issues.apache.org/jira/browse/MNG-7129)-master

I guess you're referring to https://github.com/apache/maven/tree/MNG-7129-master in your two links ... ?

**Comment by andrey-petrenko-db** [comment]
>Yes, it's similar to the commit I pointed at but done in a different way (i.e. introducing a `MojoExecutorStrategy` interface instead of a `IMojoExecutor` one).

not quite
in your PR there is no low coupling, no since extension point, you moved MojoExecutor to interface yes, but with this interface you transitively moved to public API ProjectIndex, PhaseRecorder, DependencyContext - do we really need them in public API, it is possible to keep consistency between all of them in future?
Our idea was to make change in core as small as possible to prevent breaking changes in future
Also because caching logic should not override all logic in default executor there is a code duplication in MojoExecutor and CachingMojoExecutor
Of course I think you better know how to implement this correctly than me
But when we started to think about how to move caching logic to extension we made similar changes that you have done in your PR and found out these drawbacks which I described above

All of the above is my opinion of course

> I guess you're referring to https://github.com/apache/maven/tree/MNG-7129-master in your two links ... ?
yes

**Comment by gnodet** [comment]
> > Yes, it's similar to the commit I pointed at but done in a different way (i.e. introducing a `MojoExecutorStrategy` interface instead of a `IMojoExecutor` one).
> 
> not quite in your PR there is no low coupling, no single extension point, you moved MojoExecutor to interface yes, but with this interface you transitively moved to public API ProjectIndex, PhaseRecorder, DependencyContext - do we really need them in public API, it is possible to keep consistency between all of them in future? Our idea was to make change in core as small as possible to prevent breaking changes in future Also because caching logic should not override all logic in default executor there is a code duplication in MojoExecutor and CachingMojoExecutor Of course I think you better know how to implement this correctly than me But when we started to think about how to move caching logic to extension we made similar changes that you have done in your PR and found out these drawbacks which I described above

Fair point, I think you're right that your proposal requires a lower coupling and thus is better.  

> 
> All of the above is my opinion of course
> 
> > I guess you're referring to https://github.com/apache/maven/tree/MNG-7129-master in your two links ... ?
> 
> yes

It would be nice to have a first PR to merge this change into https://github.com/apache/maven/tree/MNG-7129-master, so that we'll have a better understanding on how this new interface will be used.

**Comment by andrey-petrenko-db** [comment]
> > > Yes, it's similar to the commit I pointed at but done in a different way (i.e. introducing a `MojoExecutorStrategy` interface instead of a `IMojoExecutor` one).
> > 
> > 
> > not quite in your PR there is no low coupling, no single extension point, you moved MojoExecutor to interface yes, but with this interface you transitively moved to public API ProjectIndex, PhaseRecorder, DependencyContext - do we really need them in public API, it is possible to keep consistency between all of them in future? Our idea was to make change in core as small as possible to prevent breaking changes in future Also because caching logic should not override all logic in default executor there is a code duplication in MojoExecutor and CachingMojoExecutor Of course I think you better know how to implement this correctly than me But when we started to think about how to move caching logic to extension we made similar changes that you have done in your PR and found out these drawbacks which I described above
> 
> Fair point, I think you're right that your proposal requires a lower coupling and thus is better.
> 
> > All of the above is my opinion of course
> > > I guess you're referring to https://github.com/apache/maven/tree/MNG-7129-master in your two links ... ?
> > 
> > 
> > yes
> 
> It would be nice to have a first PR to merge this change into https://github.com/apache/maven/tree/MNG-7129-master, so that we'll have a better understanding on how this new interface will be used.

I see two options:

1. merge your PR to master then we make branch from master and reapply all changes we have done, raise PR
2. create branch from your branch reapply all changes and raise PR to it (not master) hence you can see difference

From my point of view changes from your branch should be taken anyway so the order of merge does not make much sense
@maximilian-novikov-db , @AlexanderAshitkin what do you think?

**Comment by gnodet** [comment]
> > > > Yes, it's similar to the commit I pointed at but done in a different way (i.e. introducing a `MojoExecutorStrategy` interface instead of a `IMojoExecutor` one).
> > > 
> > > 
> > > not quite in your PR there is no low coupling, no single extension point, you moved MojoExecutor to interface yes, but with this interface you transitively moved to public API ProjectIndex, PhaseRecorder, DependencyContext - do we really need them in public API, it is possible to keep consistency between all of them in future? Our idea was to make change in core as small as possible to prevent breaking changes in future Also because caching logic should not override all logic in default executor there is a code duplication in MojoExecutor and CachingMojoExecutor Of course I think you better know how to implement this correctly than me But when we started to think about how to move caching logic to extension we made similar changes that you have done in your PR and found out these drawbacks which I described above
> > 
> > 
> > Fair point, I think you're right that your proposal requires a lower coupling and thus is better.
> > > All of the above is my opinion of course
> > > > I guess you're referring to https://github.com/apache/maven/tree/MNG-7129-master in your two links ... ?
> > > 
> > > 
> > > yes
> > 
> > 
> > It would be nice to have a first PR to merge this change into https://github.com/apache/maven/tree/MNG-7129-master, so that we'll have a better understanding on how this new interface will be used.
> 
> I see two options:
> 
> 1. merge your PR to master then we make branch from master and reapply all changes we have done, raise PR
> 2. create branch from your branch reapply all changes and raise PR to it (not master) hence you can see difference
> 
> From my point of view changes from your branch should be taken anyway so the order of merge does not make much sense @maximilian-novikov-db , @AlexanderAshitkin what do you think?

I'd rather squash and merge into master the whole feature at once when it's ready, and let others review the full set of changes, so I'd rather go for #2.

**Comment by AlexanderAshitkin** [comment]
> > From my point of view changes from your branch should be taken anyway so the order of merge does not make much sense @maximilian-novikov-db , @AlexanderAshitkin what do you think?
> 
> I'd rather squash and merge into master the whole feature at once when it's ready, and let others review the full set of changes, so I'd rather go for #2.
Hi. My understanding from previous communication with Herve and Robert that minimal isolatied change in master will be better perceived than one massive drop. It could be done today and will be a step forward already. Having that merged the following work will be focused solely on extension and likely will be easier to change after decoupling from the maven-core. 

So in my understanding it is better to do this smal change. Also, Andrey has integrated the change already in upcoming PR, so he likely can create preview branch to demotstrate it if necessary

In short, i think we should merge it and move on

**Comment by gnodet** [comment]
> > > From my point of view changes from your branch should be taken anyway so the order of merge does not make much sense @maximilian-novikov-db , @AlexanderAshitkin what do you think?
> > 
> > 
> > I'd rather squash and merge into master the whole feature at once when it's ready, and let others review the full set of changes, so I'd rather go for #2.
> 
> Hi. My understanding from previous communication with Herve and Robert that minimal isolated change in master will be better perceived than one massive drop. It could be done today and will be a step forward already. Having that merged the following work will be focused solely on extension and likely will be easier to change after decoupling from the maven-core. As i understand the concern is integration itself - is this change enough or not. The answer to that is yes, it is enough. Andrey has integrated the change already in upcoming PR, so he likely can create preview branch to demonstrate it if necessary
> 
> In short, i think we should merge it and move forward
> 
> Thank you!

It's quite hard to actually review this PR without any real usage for it.  As this is kind of a backport, I think we'd need the PR for the caching system in order to understand how those new interfaces will be used.

**Comment by jvanzyl** [comment]
If this is ultimately going to be an extension, which I agree it should, can one commit be of the interfaces created that will allow the incremental capabilities in the core, and create a separate repository outside of Maven core with the extension. 

This is historically how extensions are made available: as separate artifacts. From past experience I can tell you that the implementation will change at a much higher rate than the core. I don't think it makes sense for an a new, experimental feature to be bound to the core where a change in said feature necessitates a release of the core.

**Comment by andrey-petrenko-db** [comment]
> If this is ultimately going to be an extension, which I agree it should, can one commit be of the interfaces created that will allow the incremental capabilities in the core, and create a separate repository outside of Maven core with the extension?
> 
> This is historically how extensions are made available: as separate artifacts. From past experience I can tell you that the implementation will change at a much higher rate than the core. I don't think it makes sense for an a new, experimental feature to be bound to the core where a change in said feature necessitates a release of the core.

Actually that was our plan, small as possible changes in core and then raise separate PR to extension repository
As I understood there is a concern regarding this PR because it is not clear whether it is enough or we require more changes in core
So we can create separate PR to extension repository to demonstrate how this API is used

**Comment by maximilian-novikov-db** [comment]
> If this is ultimately going to be an extension, which I agree it should, can one commit be of the interfaces created that will allow the incremental capabilities in the core, and create a separate repository outside of Maven core with the extension?
> 
> This is historically how extensions are made available: as separate artifacts. From past experience I can tell you that the implementation will change at a much higher rate than the core. I don't think it makes sense for an a new, experimental feature to be bound to the core where a change in said feature necessitates a release of the core.

this pr is exactly for allowing the caching logic to be moved to an extension

**Comment by gnodet** [comment]
> > If this is ultimately going to be an extension, which I agree it should, can one commit be of the interfaces created that will allow the incremental capabilities in the core, and create a separate repository outside of Maven core with the extension?
> > This is historically how extensions are made available: as separate artifacts. From past experience I can tell you that the implementation will change at a much higher rate than the core. I don't think it makes sense for an a new, experimental feature to be bound to the core where a change in said feature necessitates a release of the core.
> 
> Actually that was our plan, small as possible changes in core and then raise separate PR to extension repository As I understood there is a concern regarding this PR because it is not clear whether it is enough or we require more changes in core So we can create separate PR to extension repository to demonstrate how this API is used

I propose the following plan:
  * create a similar PR than this one on top of the `MNG-7129-master` branch which will modify the code to have a cleaner separation between `maven-core` and `maven-caching` and eventually become a real extension
  * merge the PR in `maven-caching`
  * merge this PR (and any additional changes in core needed) into master
  * create a back port for 3.x (if there's a consensus)
  * extract the `maven-caching` extension to its own repo (if there's a consensus)

**Comment by gnodet** [comment]
I've merge this branch to `MNG-7129-master`. We can continue working on `MNG-7129-master` and when the state is satisfying, we can come back to this PR to merge to master and extract the module to its final destination, wherever it will be...

**Comment by gnodet** [comment]
I've changed this PR to a draft until we can get #607 in a more mature state.

**Comment by gnodet** [comment]
Superseded by #607
